### PR TITLE
[FIX] clipboard: fix array formula copy

### PR DESCRIPTION
**Pull request by anhe-odoo**
## Task Description

When copying/cutting parts of the spread of an array formula, we currently copy an object with an evaluatedCell but without any cell, as there is no real cell to copy. Then, when pasting the copied values, we skip these parts as we look only for valid cell.

This commit aims to change this behaviour by pasting the evaluated cell when the cell is undefined.

## Related Task
- Task: [3541325](https://www.odoo.com/web#id=3541325&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## Review Checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 1]
```diff
@@ -75,13 +76,18 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
...
     for (let row of rowsIndex) {
-      let cellsInRow: ClipboardCell[] = [];
+      let cellsInRow: Maybe<ClipboardCell>[] = [];
       for (let col of columnsIndex) {
         const position = { col, row, sheetId };
+        const spreader = getters.getArrayFormulaSpreadingOn(position);
+        if (spreader && rowsIndex.includes(spreader.row) && columnsIndex.includes(spreader.col)) {
+          cellsInRow.push(undefined);
+          continue;
+        }
```
Comment (laa-odoo) [inline src/helpers/clipboard/clipboard_cells_state.ts:90]: if we do that, copy/past cells with their own spreader will not copy/past the style and the border of the spreading cells, no ?

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 2]
```diff
@@ -413,6 +416,10 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
         style: origin.cell.style || null,
         format: origin.cell.format,
       });
+    } else if (origin.evaluatedCell && origin.evaluatedCell.type !== "empty") {
```
Comment (laa-odoo) [inline src/helpers/clipboard/clipboard_cells_state.ts:419]: we also need to consider onlyValue and onlyFormat cases in this part

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 3]
```diff
@@ -369,58 +381,51 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
...
-      if (origin.cell.isFormula && operation === "COPY") {
-        content = this.getters.getTranslatedCellFormula(
-          sheetId,
-          col - origin.position.col,
-          row - origin.position.row,
-          origin.cell.compiledFormula,
-          origin.cell.dependencies
-        );
-      }
+    if (clipboardOption?.pasteOption === "onlyFormat") {
```
Comment (laa-odoo) [inline src/helpers/clipboard/clipboard_cells_state.ts:401]: in this case, we will dispatch an UPDATE_CELL even if origin.cell is empty

Comment (laa-odoo) [reply src/helpers/clipboard/clipboard_cells_state.ts:401]: same comment concerning onlyValue

File: src/types/misc.ts [hunk 4]
```diff
@@ -190,11 +190,12 @@ export function isMatrix(x: any): x is Matrix<any> {
 }
 
 export interface ClipboardCell {
+  content?: string;
```
Comment (LucasLefevre) [inline src/types/misc.ts:193]: Feels weird to have both `cell` and `content`. The first thing people reading this will ask themselves it what's the difference between `content` and `cell.content`. Same for `format`

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 5]
```diff
@@ -502,14 +507,22 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
...
     let htmlTable = '<table border="1" style="border-collapse:collapse">';
     for (const row of this.cells) {
       htmlTable += "<tr>";
       for (const cell of row) {
-        const cssStyle = cssPropertiesToCss(cellStyleToCss(cell.style));
+        if (!cell) {
+          continue;
+        }
+        const cssStyle = cssPropertiesToCss(
+          cellStyleToCss(this.getters.getCellComputedStyle(cell.position))
```
Comment (LucasLefevre) [inline src/helpers/clipboard/clipboard_cells_state.ts:524]: Isn't this another fix (that should be in another commit, and possibly backported to earlier versions) ? That because we don't copy the conditional formats, right ?

File: src/helpers/clipboard/clipboard_cells_state.ts [hunk 6]
```diff
@@ -369,58 +381,51 @@ export class ClipboardCellsState extends ClipboardCellsAbstractState {
...
-        this.dispatch("UPDATE_CELL", {
-          ...target,
-          style: origin.cell.style,
-          format: origin.evaluatedCell.format,
-        });
-        return;
-      }
+    if (clipboardOption?.pasteOption === "onlyValue") {
+      const locale = this.getters.getLocale();
+      const content = formatValue(origin.evaluatedCell.value ?? "", { locale });
```
Comment (LucasLefevre) [inline src/helpers/clipboard/clipboard_cells_state.ts:386]: null coalescing is useless. `value` is never undefined nor null
```suggestion
      const content = formatValue(origin.evaluatedCell.value, { locale });
```

File: src/helpers/clipboard/clipboard_cells_state.ts (hunk 3 above)
Comment (LucasLefevre) [reply src/helpers/clipboard/clipboard_cells_state.ts:401]: up

**Review by LucasLefevre — COMMENTED** [review]
can target 17.0 :wink:

File: src/helpers/clipboard/clipboard_cells_state.ts (hunk 5 above)
Comment (anhe-odoo) [reply src/helpers/clipboard/clipboard_cells_state.ts:524]: No, in fact this change is coming from the fact that, now, cell.style is not present anymore while, before, we were pushing `getCellComputedStyle(cell.position)` in the object when copying the cells. So it's not a fix, only a change caused by my other changes :)

File: src/helpers/clipboard/clipboard_cells_state.ts (hunk 3 above)
Comment (anhe-odoo) [reply src/helpers/clipboard/clipboard_cells_state.ts:401]: While I agree'd with Alexis at first time, it looks like displaying UPDATE_CELL with empty origin is a desired behavior as we have a test to ensure that copying format from an empty cell clear the format of the paste target. And same things happen when pasting the value from an empty cell to clear the content of a target cell

**Review by LucasLefevre — COMMENTED** [review]
revert those changes:
https://github.com/odoo/o-spreadsheet/pull/2473/commits/2faad1936326207d65dae533ffe4e7327a640cd2#diff-0207ac1d563650b2cfda792312504949500aa41ef99016cd31f58750249bf157L98
https://github.com/odoo/o-spreadsheet/pull/2473/commits/2faad1936326207d65dae533ffe4e7327a640cd2#diff-c71d8f0676e4628a0aac3d53bf055d95e80bcb3f582700abd3a77fd73984f892L33

It's n'importe quoi.

I think it will fix the strange system crash, since we won't even need to invalidate the cells

File: tests/clipboard/clipboard_plugin.test.ts [hunk 7]
```diff
@@ -2052,6 +2052,79 @@ describe("clipboard: pasting outside of sheet", () => {
     expect(getEvaluatedCell(model, "A1").value).toBe(8.14);
   });
 
+  test("Can copy parts of the spreaded values", () => {
```
Comment (VincentSchippefilt) [inline tests/clipboard/clipboard_plugin.test.ts:2055]: spread in the past tense is spread

File: tests/clipboard/clipboard_plugin.test.ts [hunk 8]
```diff
@@ -2052,6 +2052,79 @@ describe("clipboard: pasting outside of sheet", () => {
...
+    setCellContent(model, "A2", "2");
+    setCellContent(model, "A3", "3");
+    setCellContent(model, "B1", "=TRANSPOSE(A1:A3)");
+    copy(model, "C1:D1");
+    paste(model, "C2");
+    expect(getEvaluatedCell(model, "C2").value).toBe(2);
+    expect(getEvaluatedCell(model, "D2").value).toBe(3);
+  });
+
+  test("Cutting parts of the spreaded values will make a copy of the values", () => {
```
Comment (VincentSchippefilt) [inline tests/clipboard/clipboard_plugin.test.ts:2067]: spread in the past tense is spread

File: tests/clipboard/clipboard_plugin.test.ts [hunk 9]
```diff
@@ -2052,6 +2052,79 @@ describe("clipboard: pasting outside of sheet", () => {
...
+    cut(model, "C1:D1");
+    paste(model, "C2");
+    expect(getEvaluatedCell(model, "B1").value).toBe(1);
+    expect(getEvaluatedCell(model, "C1").value).toBe(2);
+    expect(getEvaluatedCell(model, "C2").value).toBe(2);
+    expect(getEvaluatedCell(model, "D1").value).toBe(3);
+    expect(getEvaluatedCell(model, "D2").value).toBe(3);
+  });
+
+  test("can copy and paste format only from spreaded value", () => {
```
Comment (VincentSchippefilt) [inline tests/clipboard/clipboard_plugin.test.ts:2082]: spread in the past tense is spread

File: tests/clipboard/clipboard_plugin.test.ts [hunk 10]
```diff
@@ -2052,6 +2052,79 @@ describe("clipboard: pasting outside of sheet", () => {
...
+    setCellContent(model, "A1", "=SUM(1+2)");
+
+    // formula with format seted on it
+    setCellContent(model, "A2", "=SUM(1+2)");
+    setCellFormat(model, "A2", "0%");
+
+    // formula that return value with format
+    setCellContent(model, "A3", "=DATE(2042,1,1)");
+
+    // formula that return value with format and other format seted on it
```
Comment (VincentSchippefilt) [inline tests/clipboard/clipboard_plugin.test.ts:2095]: set in the past is set

File: tests/clipboard/clipboard_plugin.test.ts [hunk 11]
```diff
@@ -2052,6 +2052,79 @@ describe("clipboard: pasting outside of sheet", () => {
...
+    setCellContent(model, "A4", "=DATE(2042,1,1)");
+    setCellFormat(model, "A4", "0%");
+
+    // formula that return value with format infered from reference
```
Comment (VincentSchippefilt) [inline tests/clipboard/clipboard_plugin.test.ts:2099]: infered --> inferred

**Comment by VincentSchippefilt** [comment]
robodoo r+

**Comment by LucasLefevre** [comment]
robodoo r+
